modifying fX and fY definition to use np.fft.fftfreq

demroz opened the pull request:
Hello,
Thank you for developing this package. Its been very useful to me for metasurface design.

Recently I've faced a fairly annoying issue with field propagation on grids with an odd number of pixel values. Specifically if you define a lens phase function on a 51 x 51 grid:

```
p = 1.0
wave = 2

nx = 51
ny = 51

D = (nx - 1) * p


F = D

x = np.arange(0,nx) - D/2
y = np.arange(0,ny) - D/2

xx,yy = np.meshgrid(x,y)

mask = np.zeros_like(xx)
mask[np.abs(xx) ** 2 + np.abs(yy) ** 2 < (D/2) **2] = 1

phase = np.mod(2*np.pi/wave * (np.sqrt(xx**2+yy**2+F**2)-F),(2*np.pi))
phase *= mask
field = np.exp(1j*phase) * mask

plt.figure()
plt.imshow(phase)
plt.show()
```
![phase](https://github.com/user-attachments/assets/84beeef5-ac46-4c1a-8a63-493910017b6a)

then propagate it to the focal spot with bandlimit = False, pad = False, I get the following fields

![fafield image](https://github.com/user-attachments/assets/e9cd0119-e494-47eb-9413-21a612c3ca1d)

the left ASM farfield I got from the old version of the package, and the right is the field I got with the changes I added to this PR. Since the phase is radially symmetric around x,y=0, i expect the maximum to occur at x,y=0 at the focal length. With the old version I get the maximum at x=-1, y = -1 at the focal length. See image below:

![slices](https://github.com/user-attachments/assets/f999b224-3d70-4e08-8b57-e25416a140ee)

It seems that the way the fourier frequencies are defined are inconsistent between even/odd grids, and np.fft.fftfreq should be used to generate the frequencies. Also, setting the grid diameters Dx/Dy as 
``
Dy, Dx = (d1[0] * float(Ny_pad), d1[1] * float(Nx_pad))
``
gives values that are 1 pixel larger than expected. The correct Dx, Dy should be given by 
```
Dy, Dx = (d1[0] * float(Ny_pad - 1), d1[1] * float(Nx_pad - 1))
```

Thank you

Commit:
modified fX and fY generation to use np.fft.fftfreq

Commit:
fixed inconsistent diameter

Commit:
updating

Commit:
weird inconsistency between compuiters

roflmaostc commented:
I think to get this merged, it's better to sync your main branch to the one of this repository.
There is too many changes for a reasonable review.

But I agree, fftfreq is preferred over arange!

Commit:
more fixes

Commit:
io.py remove

Commit:
fixes

Commit:
grid update to fraunhoffer.py

Commit:
updates to fresnel diffraction

Commit:
fresnel diffraction updates

ebezzam commented on waveprop/fraunhofer.py:
```diff
@@ -30,7 +30,9 @@ def fraunhofer(u_in, wv, d1, dz):
 
     # frequencies of propagating waves
-    fx, fy = sample_points(N=[Ny, Nx], delta=[(1 / Ny / d1[0]), (1 / Nx / d1[1])])
+    # fx, fy = # sample_points(N=[Ny, Nx], delta=[(1 / Ny / d1[0]), (1 / Nx / d1[1])])
```
If we are sure about changes we can delete commented lines of old code (if we want to go back we can check git history).

ebezzam commented on waveprop/fresnel.py:
```diff
@@ -32,14 +32,15 @@ def fresnel_one_step(u_in, wv, d1, dz):
+    xx1, yy1 = np.meshgrid(x1,y1)
     x2, y2 = sample_points(N=[Ny, Nx], delta=[1 / Ny / d1[0] * wv * dz, 1 / Nx / d1[1] * wv * dz])
-
+    xx2, yy2 = np.meshgrid(x2,y2)
```
is the mesh grid needed? from what I remember x1 and y2 and row and column vectors so that broadcasting avoids the need of meshgrid.

ebezzam commented on waveprop/fresnel.py:
```diff
@@ -147,25 +148,32 @@ def fresnel_conv(u_in, wv, d1, dz, device=None, dtype=None, d2=None, pad=True):
     x1, y1 = sample_points(N=N, delta=d1)
-    r1sq = x1**2 + y1**2
+    xx1, yy1 = np.meshgrid(x1,y1)
+    r1sq = xx1**2 + yy1**2
```
similar as above, is meshgrid really needed?

ebezzam commented on waveprop/fresnel.py:
```diff
@@ -147,25 +148,32 @@ def fresnel_conv(u_in, wv, d1, dz, device=None, dtype=None, d2=None, pad=True):
+    df1[0] = fX[1] - fX[0]
+    df1[1] = fY[1] - fY[0]
+    fXX,fYY = np.meshgrid(fX,fY)
+    fsq = fXX**2 + fYY**2
```
ok this is the main of the PR 🙂 

how about making a helper function like `sample_points`? to avoid duplication of code. Something like:

```python
def sample_freq(N, delta, pytorch=False):
    """
    Return frequency sampling.

    Parameters
    ----------
    N : int or list
        Number of sample points
    delta: int or float or list
        Sampling period along x-dimension and (optionally) y-dimension [m].
    """
   if isinstance(N, int):
      N = [N, N]
   assert len(N) == 2
   if isinstance(delta, float) or isinstance(delta, int):
      delta = [delta, delta]
   fx = np.fft.fftshift(np.fft.fftfreq(N[0], delta[0]))[np.newaxis, :]
   fy = np.fft.fftshift(np.fft.fftfreq(N[1], delta[1]))[:, np.newaxis]
   df = np.array([fX[1] - fX[0], fY[1] - fY[0]])
   if pytorch:
      fx = torch.tensor(fx)
      fy = torch.tensor(fy)
   return fx, fy, df
```

So that in this part of the code (and others) we have the following (which also doesn't need meshgrid)
```python
# source spatial frequencies
fx, fy, df = sample_freq(N, delta, pytorch=pytorch)
fsq = fx**2 + fy**2
```

ebezzam commented on waveprop/fresnel.py:
```diff
@@ -147,25 +148,32 @@ def fresnel_conv(u_in, wv, d1, dz, device=None, dtype=None, d2=None, pad=True):
     x2, y2 = sample_points(N=N, delta=d2)
-    r2sq = x2**2 + y2**2
+    xx2, yy2 = np.meshgrid(x2,y2)
+    r2sq = xx2**2 + yy2**2
```
is meshgrid needed?

demroz replied:
Was breaking one of my scripts because sample_points doesn't return 1xN, Nx1 x and ys, but N, N, shaped arrays. I'll update

demroz replied:
Actually seems I broke it by editing how x,y gets generated in sample_points. I'll fix it

ebezzam commented on waveprop/fresnel.py:
```diff
@@ -180,7 +188,8 @@ def fresnel_conv(u_in, wv, d1, dz, device=None, dtype=None, d2=None, pad=True):
         u_out = crop(u_out, shape=N_orig, topleft=(int(N_orig[0] // 2), int(N_orig[1] // 2)))
-
+        x2 = crop(xx2, shape=N_orig, topleft=(int(N_orig[0] // 2), int(N_orig[1] // 2)))[0,:]
+        y2 = crop(yy2, shape=N_orig, topleft=(int(N_orig[0] // 2), int(N_orig[1] // 2)))[:,0]
```
can be removed if not using meshgrid?

demroz replied:
Need to be modified to match the size of u_out aswell

ebezzam commented on waveprop/rs.py:
```diff
@@ -274,13 +274,24 @@ def angular_spectrum_np(
     # size of the padded field
     Ny_pad, Nx_pad = u_in_pad.shape
-    Dy, Dx = (d1[0] * float(Ny_pad), d1[1] * float(Nx_pad))
+    Dy, Dx = (d1[0] * float(Ny_pad - 1), d1[1] * float(Nx_pad - 1))
```
is this applicable when Ny_pad, Nx_pad are even?

demroz replied:
Yes, this is about the symmetry of the grid about 0

ebezzam commented on waveprop/rs.py:
```diff
@@ -274,13 +274,24 @@ def angular_spectrum_np(
+    # dfX = 1.0 / Dx
+    # dfY = 1.0 / Dy
+    # fX_old = np.arange(-Nx_pad / 2 , Nx_pad / 2)[np.newaxis, :] * dfX
+    # fY_old = np.arange(-Ny_pad / 2 , Ny_pad / 2)[np.newaxis, :] * dfY
```
can remove commented old code

ebezzam commented on waveprop/rs.py:
```diff
@@ -274,13 +274,24 @@ def angular_spectrum_np(
+    fX = np.fft.fftshift(np.fft.fftfreq(Nx_pad, d = d1[0]))[np.newaxis,:]
+    fY = np.fft.fftshift(np.fft.fftfreq(Ny_pad, d = d1[1]))[:,np.newaxis] 
+    dfX = fX[0,1] - fX[0,0]
+    dfY = fY[1,0] - fY[0,0]
```
use helper function like above

ebezzam commented on waveprop/rs.py:
```diff
@@ -538,14 +549,18 @@ def angular_spectrum(
     # size of the padded field
     Ny_pad, Nx_pad = u_in_pad.shape
-    Dy, Dx = (d1[0] * float(Ny_pad), d1[1] * float(Nx_pad))
+    Dy, Dx = (d1[0] * float(Ny_pad - 1 ), d1[1] * float(Nx_pad - 1))
```
also applicable if Ny_pad and Nx_pad are even?

demroz replied:
Yes. the Dx/Dy calculations aren't dependant on Nx/Ny being even or odd. Its more about the symmetry of the grid

ebezzam commented on waveprop/rs.py:
```diff
@@ -538,14 +549,18 @@ def angular_spectrum(
+    # dfX = 1.0 / Dx
+    # dfY = 1.0 / Dy
+    # fX_old = np.arange(-Nx_pad / 2 , Nx_pad / 2)[np.newaxis, :] * dfX
+    # fY_old = np.arange(-Ny_pad / 2 , Ny_pad / 2)[np.newaxis, :] * dfY
```
remove old commented code

ebezzam commented on waveprop/rs.py:
```diff
@@ -538,14 +549,18 @@ def angular_spectrum(
+    fX = np.fft.fftshift(np.fft.fftfreq(Nx_pad, d = d1[0]))[np.newaxis,:]
+    fY = np.fft.fftshift(np.fft.fftfreq(Ny_pad, d = d1[1]))[:,np.newaxis] 
+    dfX = fX[0,1] - fX[0,0]
+    dfY = fY[1,0] - fY[0,0]
```
use helper function

ebezzam commented on waveprop/rs.py:
```diff
@@ -813,13 +828,18 @@ def _form_transfer_function(
+    # dfX = 1.0 / Dx
+    # dfY = 1.0 / Dy
+    # fX_old = np.arange(-Nx_pad / 2 , Nx_pad / 2)[np.newaxis, :] * dfX
+    # fY_old = np.arange(-Ny_pad / 2 , Ny_pad / 2)[np.newaxis, :] * dfY
```
remove old commented code

ebezzam commented on waveprop/rs.py:
```diff
@@ -813,13 +828,18 @@ def _form_transfer_function(
+    fX = np.fft.fftshift(np.fft.fftfreq(Nx_pad, d = d1[0]))[np.newaxis,:]
+    fY = np.fft.fftshift(np.fft.fftfreq(Ny_pad, d = d1[1]))[:,np.newaxis] 
+    dfX = fX[0,1] - fX[0,0]
+    dfY = fY[1,0] - fY[0,0]
```
use helper function

ebezzam commented on waveprop/rs.py:
```diff
@@ -916,7 +936,7 @@ def _bandpass(H, fX, fY, Sx, Sy, x0, y0, z0, wv):
     :return:
     """
-    du = 1 / (Sx) # not divided by 2 like in 17 since field already padded
+    du = 1 / (2 * Sx)
```
are you sure about multiplying by 2? Sx and Sy are already padded when given as an input: https://github.com/ebezzam/waveprop/blob/84c3e7dd50c7193e4c069a7581740ecbf33a3167/waveprop/rs.py#L864

I agree that naming could have been better. Dx and Dy might have been better (D for destination plane)

ebezzam commented on waveprop/rs.py:
```diff
@@ -929,7 +949,7 @@ def _bandpass(H, fX, fY, Sx, Sy, x0, y0, z0, wv):
         u_width = u_limit_p + u_limit_n
 
-    dv = 1 / (Sy) # not divided by 2 like in 18 since field already padded
+    dv = 1 / (2 * Sy)
```
same here, should it multiplied by 2 if Sy is already padded?

demroz replied:
Reverted. I think I was just doing some testing on my end and forget to change it back

ebezzam commented on waveprop/rs.py:
```diff
@@ -916,7 +936,7 @@ def _bandpass(H, fX, fY, Sx, Sy, x0, y0, z0, wv):
+    du = 1 / (2 * Sx)
     u_limit_p = ((x0 + 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
     u_limit_n = ((x0 - 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
     if Sx < x0:
```
but actually this should rather be ``Sx / 2 < x0`` since Sx is padded (and same with Sy/2) below.

ebezzam commented on waveprop/rs.py:
```diff
@@ -929,7 +949,7 @@ def _bandpass(H, fX, fY, Sx, Sy, x0, y0, z0, wv):
+    dv = 1 / (2 * Sy)
     v_limit_p = ((y0 + 1 / (2 * dv)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
     v_limit_n = ((y0 - 1 / (2 * dv)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
     if Sy < y0:
```
I think should rather be `if (Sy/2) < y0` if Sy is already padded

ebezzam commented on waveprop/rs_old.py:
we can delete this file instead of keeping it as Git tracks past versions

ebezzam commented on waveprop/simulation.py:
```diff
@@ -40,8 +40,6 @@ def __init__(
         is_torch=False,
         quantize=True,
         return_float=True,
-        vertical_shift=None,
```
why did you remove these shifts?

demroz replied:
I'm not sure how any of these got removed. I think it was an accident and i didn't change it

ebezzam commented on waveprop/util.py:
```diff
@@ -102,13 +102,23 @@ def sample_points(N, delta, shift=0, pytorch=False):
         delta = torch.tensor(delta)
         shift = torch.tensor(shift)
-        x = torch.arange(-N[1] / 2, N[1] / 2) * delta[1] + shift[1]
+        Dx = (N[1] - 1) * delta[1]
```
is this also valid if N is even?

demroz replied:
Yes if we want the unshifted grid to be symmetric about 0, i.e. min(abs(x)) = max(abs(x))

ebezzam commented on waveprop/util.py:
```diff
@@ -102,13 +102,23 @@ def sample_points(N, delta, shift=0, pytorch=False):
         x = torch.unsqueeze(x, 0)
-        y = torch.arange(-N[0] / 2, N[0] / 2) * delta[0] + shift[0]
         y = torch.unsqueeze(y, 1)
+        #x = torch.arange(-N[1] / 2, N[1] / 2) * delta[1] + shift[1]
```
can remove commented old code

ebezzam commented on waveprop/util.py:
```diff
@@ -432,18 +442,23 @@ def zero_pad(u_in, pad=None):
-            y_pad_edge + 1 if Ny % 2 else y_pad_edge,
-            y_pad_edge,
-        )
+        # pad_width = (
```
are you sure that extra sample is needed if Nx, Ny are odd?

ebezzam reviewed (changes requested):
Thanks @demroz and @mzhelyez for your PR, and @roflmaostc for jumping in!

Sorry for the delay, as I'm in the thesis writing / wrapping up stage! Also I've been moving my implementations to [chromatix](https://github.com/chromatix-team/chromatix/), so it might be interesting for you try your setup with that library, which is more stable and:
- [Already does](https://github.com/chromatix-team/chromatix/blob/820f02ca45a1bd0f2f8fc9122294a2fc93fcf1b3/src/chromatix/field.py#L117) your suggestion for frequency sampling, which I agree is needed.
- Has a cleaner implementation of [bandlimiting](https://github.com/chromatix-team/chromatix/blob/820f02ca45a1bd0f2f8fc9122294a2fc93fcf1b3/src/chromatix/functional/propagation.py#L400), for which I also see you made suggestions.

`waveprop` has been a rather dirty dump of implementations that helped me familiarize with free-space propagation and to use for my other project. But I'm happy it's had some use for you 🙂 

But if you would like to stick with `waveprop` I have requested some changes. There's a lot of small comments, but the main ideas:
- Synchronizing your fork again. I've updated requirements and added automated tests (which you can also run locally with `pytest`). From your fork, I noticed some tests were failing when I ran `pytest`  (also because I had to update to newer numpy conventions).
- Defining a utility function for frequency sampling to avoid code duplication
- Relying on broadcasting to avoid meshgrid.
- Deleting old code (instead of commenting) as git will keep track of previous versions if we really want to go back to them
- Making sure padding/sizes work for both odd and even. Perhaps adding a test for that [here](https://github.com/ebezzam/waveprop/blob/master/tests/test_pytorch_prop_function.py).

Hope that helps, and thanks again for your PR 🚀

demroz committed:
Merge branch 'ebezzam:master' into master

Commit:
test branch

Commit:
code review updates

Commit:
fixed sample points to pass tests

Commit:
added even/odd propagation test

demroz commented:
Hello @ebezzam and @roflmaostc 

I've looked over the suggestions you've made in your review. I believe I've fixed all the points of contention, and added tests for even/odd propagation tests. Thank you for your time.

Commit:
removed newaxis from fresnel coordinate definitions

Commit:
change rs

Commit:
fix waveprop
